### DISCUSSION TITLE
Need help to make rust-analyzer work

### DISCUSSION THREAD

**Discussion by Hidan0** [esp-rs/espup#244] [Q&A] [answered]
I'm trying to setup my development environment for an ESP32, but I'm having issues with rust-analyzer. I've installed the toolchain with `espup install` and sourced the envs. 

When my lsp starts it crashed with this error:
```
error: 'rust-analyzer' is not installed for the toolchain 'esp'
note: this is a custom toolchain, which cannot use `rustup component add`
help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`
```

I'm using Neovim and it works fine with the stable toolchain.

These are the RA settings:
```lua
settings = {
      ["rust-analyzer"] = {
        checkOnSave = true,
        check = {
          command = "clippy",
        },
      },
    },
```

Any idea on how to fix/work around this problem?

**Comment by SergioGasquez** [comment]
Hi! I can't really help since I do not use Neovim, but I don't have any issues with RA in VS Code. `espup` install the artifacts that [rust-build](https://github.com/esp-rs/rust-build) generates after building [rust](https://github.com/esp-rs/rust).

Here are a few similar issues that might be useful to you:
- https://github.com/esp-rs/esp-template/issues/58
- https://github.com/esp-rs/esp-idf-template/issues/74

**Comment by Hidan0** [comment, marked as answer]
With some help I found a solution: forcing nvim to use RA from [Mason](https://github.com/williamboman/mason.nvim) (a plugin to manage LSP servers).

I think it was using RA from the stable toolchain before and that was causing the crash. After removing RA component there still was a link to rustup (`/usr/bin/rust-analyzer -> /usr/bin/rustup`) --I don't know why--. Deleting the link "forced" nvim to use RA from Mason, and "fixed" the problem.

**Comment by x1unix** [comment]
**2026 Update:** The recent version of rust-analyzer isn't compatible, therefore autocomplete might not work.

```
2026-08-02T16:55:22.287216-04:00  WARN `cargo metadata` failed and returning succeeded result with `--no-deps` error=`cargo metadata` exited with an error: error: unexpected argument '--lockfile-path' found

  tip: a similar argument exists: '--locked'

Usage: cargo metadata --format-version <VERSION> --manifest-path <PATH> --filter-platform <TRIPLE> --locked

For more information, try '--help'.


Stack backtrace:
   0: <std::backtrace::Backtrace>::create
   1: <anyhow::Error as core::convert::From<cargo_metadata::errors::Error>>::from
   2: <project_model::cargo_workspace::FetchMetadata>::exec::{closure#2}
   3: <project_model::cargo_workspace::FetchMetadata>::exec
   4: <project_model::sysroot::Sysroot>::load_workspace
   5: std::sys::backtrace::__rust_begin_short_backtrace::<<project_model::workspace::ProjectWorkspace>::load_cargo::{closure#1}::{closure#4}, core::option::Option<project_model::sysroot::RustLibSrcWorkspace>>
   6: <std::thread::lifecycle::spawn_unchecked<<project_model::workspace::ProjectWorkspace>::load_cargo::{closure#1}::{closure#4}, core::option::Option<project_model::sysroot::RustLibSrcWorkspace>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
   7: <std::sys::thread::unix::Thread>::new::thread_start
   8: __pthread_cond_wait
```

I was able to "hack" that by using an old version of RA which my package manager shipped.

**Reply by SergioGasquez** [reply]
@x1unix 1.96 works - or  work arounds in https://github.com/rust-lang/rust-analyzer/issues/22371

**Reply by x1unix** [reply]
@SergioGasquez thanks for the tip!
